packages/bsync/src/purchases/purchasesClient.ts

```diff
@@ -0,0 +1,272 @@
+import { Subscription } from '../proto/bsync_pb'
+import { RcSubscription, RevenueCatClient } from './revenueCatClient'
+
+type PlatformId = 'web' | 'ios' | 'android'
+
+type SubscriptionGroupId = 'core'
+
+type SubscriptionOfferingId = 'core:monthly' | 'core:annual'
+
+const assertSubscriptionGroup: (
+  group: string,
+) => asserts group is SubscriptionGroupId = (group: string) => {
+  if (['core'].includes(group)) {
+    return
+  }
+  throw new Error(`invalid subscription group: '${group}'`)
+}
+
+const assertPlatform: (platform: string) => asserts platform is PlatformId = (
+  platform: string,
+) => {
+  if (['web', 'ios', 'android'].includes(platform)) {
+    return
+  }
+  throw new Error(`invalid platform: '${platform}'`)
+}
+
+export type PurchasesClientOpts = {
+  revenueCatV1ApiKey: string
+  revenueCatV1ApiUrl: string
+  revenueCatWebhookAuthorization: string
+  stripePriceIdMonthly: string
+  stripePriceIdAnnual: string
+  stripeProductIdMonthly: string
+  stripeProductIdAnnual: string
+}
+
+const mapStoreToPlatform = (store: string): PlatformId | undefined => {
+  switch (store) {
+    case 'stripe':
+      return 'web'
+    case 'app_store':
+      return 'ios'
+    case 'play_store':
+      return 'android'
+    default:
+      return undefined
+  }
+}
+
+export class PurchasesClient {
+  private revenueCatClient: RevenueCatClient
+  private STRIPE_PRODUCTS: {
+    [subscriptionOfferingId in SubscriptionOfferingId]: string
+  }
+
+  /**
+   * All of our purchaseable product IDs for all platforms.
+   */
+  private PRODUCTS: {
+    [platform in PlatformId]: {
+      [subscriptionOfferingId in SubscriptionOfferingId]: string
+    }
+  }
+
+  /**
+   * Manual groupings of products into "Subscription Groups", mimicking the way
+   * Apple does it.
+   */
+  private GROUPS: {
+    [group in SubscriptionGroupId]: {
+      [platform in PlatformId]: {
+        id: SubscriptionOfferingId
+        product: string
+      }[]
+    }
+  }
+
+  constructor(opts: PurchasesClientOpts) {
+    this.revenueCatClient = new RevenueCatClient({
+      v1ApiKey: opts.revenueCatV1ApiKey,
+      v1ApiUrl: opts.revenueCatV1ApiUrl,
+      webhookAuthorization: opts.revenueCatWebhookAuthorization,
+    })
+
+    this.STRIPE_PRODUCTS = {
+      'core:monthly': opts.stripeProductIdMonthly,
+      'core:annual': opts.stripeProductIdAnnual,
+    }
```

Comment:
I bet we can get some tighter types on these with some `as const`.
```suggestion
    this.STRIPE_PRODUCTS = {
      'core:monthly': opts.stripeProductIdMonthly,
      'core:annual': opts.stripeProductIdAnnual,
    } as const
```

Reply:
Ah I see now you define the tight types above— that works too!